File: BaekJoon/Gold/5639/C++/5639.cpp
// no.5639: 이진 검색 트리 (G4)

#include <cstdio>
using namespace std;

void postorder(int);
void insert(int, int);
int tree[1'000'000][2];

int main() {
    int n;
    int root = -1;
    while(scanf("%d", &n) == 1) {
        if(root == -1) {
            root = n;
            continue;
        }
        insert(root, n);
    }
    postorder(root);
    return 0;
}

void insert(int root, int val) {
    if(val < root) {
        if(tree[root][0] == 0)
            tree[root][0] = val;
        else
            insert(tree[root][0], val);
    }
    else {
        if(tree[root][1] == 0)
            tree[root][1] = val;
        else
            insert(tree[root][1], val);

    }
}

void postorder(int root) {
    if(tree[root][0] != 0)
        postorder(tree[root][0]);
    if(tree[root][1] != 0)
        postorder(tree[root][1]);
    printf("%d\n", root);
}

File: BaekJoon/Gold/5639/Python/5639.py
# no. 5639: 이진 검색 트리 (Gold IV)
# 이진 검색 트리를 전위 순회한 결과가 주어졌을때 -> 후위 순회한 결과를 구해라
# 트리의 삽입까지만 구현, 후위 순회하면됨

from sys import stdin
from sys import setrecursionlimit
input = stdin.readline
setrecursionlimit(10**5)

tree = {}
root = None
def insert(val, node=None):
    if node == None:
        global root
        root = val
        tree[root] = []
        return
    if val > node:
        if not tree[node]:
            tree[node] = [None, val]
            tree[val] = []
        elif tree[node][1] == None:
            tree[node][1] = val
            tree[val] = []
        else:
            insert(val, tree[node][1])
    else:
        if not tree[node]:
            tree[node] = [val, None]
            tree[val] = []
        elif tree[node][0] == None:
            tree[node][0] = val
            tree[val] = []
        else:
            insert(val, tree[node][0])

def traverse(node):
    if tree[node] and tree[node][0] != None:
        traverse(tree[node][0])
    if tree[node] and tree[node][1] != None:
        traverse(tree[node][1])
    print(node)
    
num = input().rstrip()
while num:
    insert(int(num), root)
    num = input().rstrip()

traverse(root)